docs/rfcs/0003-ecommerce-models.md

```diff
@@ -0,0 +1,136 @@
+it is `created`. Since a user may check out and then decide not to complete the purchase in
+CyberSource, orders may remain in the `created` state indefinitely.
+
+When a user checks out, they will click a button indicating this in the UI. The `Order` is created
+and attached to the user's basket. The basket is made not editable and fields are filled out
+with price data. This information is then put into a dict, with most fields signed so the user
+cannot tamper with the information.
+
+The dict is returned in the POST response to the frontend, then the user POSTs to CyberSource.
+If the user successfully completes a payment CyberSource will POST back to our order fulfillment
+API which will record the response in `Receipt` and change the `Order` status to `fulfilled`. If
+the status in the order POST back is not a success, `Order.status` will be `failed`.
+ 
+At some future time we will implement refund handling and this will set the `Order` status
+to `refunded`.
+
+#### Coupon
+
+A `Coupon` represents a coupon code and stores information about what kind of discount and
+to what product it can be applied. A `Coupon` may be used multiple times by multiple people,
+depending on the type of coupon.
+
+Once a `Coupon` is created it should **not** be edited. Unlike a `Product`, it is not expected
+that admins will edit fields of a `Coupon`. This is because a `Coupon` will be linked to from
+a `Basket` via `CouponBasket` and it provides information about the discount given to a user
+when they completed an order. `CouponBasket` and `CouponProduct` should also not be edited for
+the same reason.
+
+If an admin needs to update `Coupon` information it's recommended to set `enabled` to False on
+the existing coupons, then create new `Coupon`s to use instead.
+
+It has these fields:
+ - `coupon_code` - a unique string to be used by the end user
+ - `max_users` - Maximum number of users who can user the coupon
+ - `max_redemptions_per_user` - Maximum number of coupon redemptions per user. 
+ - `amount_type` - percent or fixed price discount
+ - `amount` - The amount of discount provided, depends on `amount_type`.
+ - `enabled` - If false, the coupon is not redeemable
+ - `expiration_date` - The latest time a coupon can be redeemed
+ - `activation_date` - The earliest time a coupon can be redeemed
+ 
+#### CouponProduct
+
+A `CouponProduct` is a link from a `Coupon` to `Product` describing which products the `Coupon`
+applies to. Unlike in MicroMasters a `Coupon` to a `Program` will **not** apply to `Course`
+or `CourseRun` which have the `Program` as a parent. Instead links to each of the items to be
+purchased must be explicitly created.
+
+#### CouponBasket
+
+A `CouponBasket` is a link from a `Coupon` to `Basket` describing the intent of a user to
+redeem the coupon on checkout (or that they already did). For right now there should be only
+one `CouponBasket` per `Basket` because only one coupon is allowed per purchase.
```

Comment:
so a basket isn't necessarily a full check-out - do we need to store on the `CouponBasket` model (which I might suggest calling `CouponUsage` or something) information about whether the user finished the whole check-out? Just wondering because if for instance we need to count the number of times a coupon has already been used we couldn't just count the rows in this table, we'd have to join to the basket and then the order table to see if the order actually completed. 

Alternatively maybe this table could be called a `CouponSelection`, since it represents the user's intent to use it by adding it to their basket, and then after a successful check-out a `CouponUsage` record could be created that joined `Order` to `Coupon`, and the `CouponUsage` would be the record to count to find the number of uses of a given coupon. Not sure what's best! haha.